#18 #14 Ft update product

Shumba250 opened the pull request:
#### What does this PR do?

- a seller should update a specific item in a collection
- only the owner of the product is authorized to update a specific item
- a seller can update an existing product
- all inputs are validated 
- appropriate error messages are displayed upon the failure or success of the request
#### Description of Task to be completed

- a seller is authorized to update a product 
- all inputs are validated
- The appropriate 4XX response message is sent to the requesting client describing the validation error
#### How should this be manually tested?

- clone the repo
- navigate to the ft-update-product branch
- install packages by typing npm install
- include Cloudinary variables in .env cloud_name = "" ,  api_key = "", api_secret = ""
- start the server by typing npm run dev
- open your browser to use swagger doc by navigating to localhost/5050/docs
- create vendor a role and permissions
- register a vendor or if you have one created, log the vendor in
- after completing the 2FA authentication, the vendor with be logged in automatically
- on swagger doc, use create product endpoint and only a minimum of 4 images are allowed and create a product
- after creating a product, use update product endpoint and use the product id to update a product

wakaflorien commented on src/controllers/userController.js:
```diff
@@ -86,6 +85,7 @@ class UserController {
       return res.status(400).json({ status: "error", error: error.message });
```
- for consitency and uniformity purposes use something like this on responses instead:

`res.status(400).json({ status: "fail", error: error.message });`

wakaflorien commented on src/controllers/userController.js:
```diff
@@ -109,6 +109,7 @@ class UserController {
       return res.status(500).json({ status: "error", error: error.message });
```
- same here

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    if (!product || product == null) {
+      return res
+        .status(404)
+        .json({ status: 404, message: "product does not exist" });
```
- avoid redudant status codes do something like thsi instead:

`res
        .status(404)
        .json({ status: "fali", message: "product does not exist" });`

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    req.product = product;
+    next();
+  } catch (error) {
+    return res.status(500).json({ status: 500, error: "server error" });
```
### same here

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    });
+    if (product) {
+      return res
+        .status(409)
```
- ### do the same here

wakaflorien commented on src/middlewares/productExists.js:
```diff
@@ -0,0 +1,37 @@
+    req.product = product;
+    next();
+  } catch (error) {
+    return res.status(500).json({ status: 500, error: error.message });
```
- ### again

wakaflorien committed:
Merge pull request #29 from atlp-rwanda/ft-create-product

#14 Ft create product

Shumba250 committed:
ft-update-product
update a particular product
request body is validated
A json object with a status: 4xx, error: the appropriate error message should be sent
